### PR TITLE
St fix round4

### PR DESCRIPTION
**Pull request by glevkovich**
1. st-ctl: misc changes
2. RangeValidationTree: minor refactor, add debugging info
3. RVBManager: refactor assert, fix bug
4. bcstatetransfer_tests: fix bug in DataGenerator::generateBlocks
5. BCStateTran: add more stats, remove ItemDataMsg timestamp correction
6. kv_blockchain: reword comment
7. TesterReplica: set gettingMissingBlocksSummaryWindowSize to 60
8. test_skvbc_state_transfer: adapt to recent changes
9. BCStateTran: convertMillisecToReadableStr pad with zeroes when needed
10. throughput: refactor. Adapt users accordingly

### CODE REVIEW

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 1]
```diff
@@ -184,7 +184,7 @@ BCStateTran::Metrics BCStateTran::createRegisterMetrics() {
       metrics_component_.RegisterCounter("src_handle_FetchBlocks_msg"),
       metrics_component_.RegisterCounter("src_handle_FetchResPages_msg"),
       metrics_component_.RegisterCounter("dst_handle_RejectFetching_msg"),
```
Comment (mdarade) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp]: Maybe other counters can be corrected at some other time.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 2]
```diff
@@ -746,25 +761,31 @@ std::string BCStateTran::convertMillisecToReadableStr(uint64_t ms) const {
   auto days = n;
 
   std::string str;
+  bool shouldPad = false;
   if (days) {
-    str += std::to_string(days) + ":";
-    legend += "D:";
+    str += padZeroes(std::to_string(days), 2, shouldPad) + ":";
+    legend += "DD:";
```
Comment (mdarade) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp]: Should legend & shouldPad flag be passed & processed in existing/new lambda as well?

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 3]
```diff
@@ -1759,7 +1771,7 @@ bool BCStateTran::onMessage(const FetchBlocksMsg *m, uint32_t msgLen, uint16_t r
                                             m->rvbGroupId,
                                             preFetchBlockId));
   ++sourceBatchCounter_;
-  DurationTracker<std::chrono::microseconds> timeWaitedForCtx;  // TODO(GL) - remove when unneeded
+  DurationTracker<std::chrono::microseconds> timeWaitedForCtx("timeWaitedForCtx");  // TODO(GL) - remove when unneeded
```
Comment (mdarade) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:1765]: I don't think std::chrono needs to be specified explicitly at all places as TimeUtils.hpp is already included here and it has relevant declarations.

File: util/include/throughput.hpp [hunk 4]
```diff
@@ -17,56 +17,78 @@
...
     }
+    LOG_INFO(GL, KVLOG(name_, running_, total_duration_, ret, do_reset));
     return ret;
   };
 
- private:
+ protected:
   uint64_t total_duration_ = 0;
-  std::chrono::time_point<std::chrono::steady_clock> start_time_;
+  using TimePoint = std::chrono::time_point<std::chrono::steady_clock>;
```
Comment (mdarade) [inline util/include/throughput.hpp:106]: I think all time related declarations should go at corresponding header file.

**Review by mdarade — APPROVED** [review]
Thanks Gil.
Few minor comments.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 2 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp]: That's a good idea, I'll make the changes. Thanks!

Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp]: shouldPad doesnt need ot be passed, just captured.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 5]
```diff
@@ -725,6 +730,16 @@ std::string BCStateTran::convertMillisecToReadableStr(uint64_t ms) const {
   }
   string legend;
   std::ostringstream oss;
+  auto padZeroes = [](std::string &&str, size_t expectedSize, bool shouldPad) {
+    ConcordAssertGE(expectedSize, str.size());
```
Comment (StefanPalashev) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp]: Usually padding functions return the same string that was passed to the function when the `expectedSize` is less than the one of the string itself.

File: scripts/st-ctl.sh [hunk 6]
```diff
@@ -458,20 +462,41 @@ if [ ! -e "/keep" ] || [ ! -e "/mnt/data/keep/" ]; then
...
+  access_key=$(cat /config/concord/config-local/deployment.config | grep -m1 s3-access-key| cut -d ' ' -f 6)
+  ip=$(echo ${ip} | xargs)
+  port=$(echo ${port} | xargs)
+  secret_key=$(echo ${secret_key} | xargs)
+  access_key=$(echo ${access_key} | xargs)
+  echo "Adding alias: ./mc alias set min http://${ip}:${port} ${access_key} ${secret_key}"
+  ${minio_command_line_tool_path} alias set ${minio_alias_name}/ http://${ip}:${port} ${access_key} ${secret_key}
 fi
 
+echo "Done Installing tools, please log in and out"
```
Comment (StefanPalashev) [inline scripts/st-ctl.sh]: Maybe the second part of the message can be rewritten as `please relog`, since the user is already logged in.

**Review by StefanPalashev — APPROVED** [review]
LGTM, thanks @glevkovich.

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 1 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp]: I will revert the change.

File: util/include/throughput.hpp (hunk 4 above)
Comment (glevkovich) [reply util/include/throughput.hpp:106]: Actually, this is a very bad thing to have using-drectives in an hpp.
So much was written about it in numerous blogs and websites.
In practice we should use it only in class or function scope.
Relaying on a using-directive of an included header is not a good idea as well.
At most, we should declare it on the top of our cpp - even in that case it is not a good thing to do (user directive at cpp file scope).
To summarize from https://www.learncpp.com/cpp-tutorial/using-declarations-and-using-directives:
"Prefer explicit namespaces over using statements. Avoid using directives whenever possible. Using declarations are okay to use inside blocks."

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 5 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp]: I'm not returniung here anymore, so ignoring this comment (pushing a fix soon which wad done on another comment).

File: bftengine/src/bcstatetransfer/BCStateTran.cpp (hunk 3 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/BCStateTran.cpp:1765]: 1. relaying on using directive of other hpp is a bad thing. If we decide we need it, you must put it in the head of your cpp.
2. using directive in hpp file is a bad thing that can cause callisions and redibility issues in the future, and even bugs.
2. https://www.learncpp.com/cpp-tutorial/using-declarations-and-using-directives/ - using directives should be used in cpp, better inside functions. if some name is not used much, I dont think we should immidiately declare a using directive, also - I tink it is belong to refactoring area.
